pep-9999.rst

```diff
@@ -811,16 +811,73 @@ to weak references.
 Hashing
 =======
 
-TODO
-https://discuss.python.org/t/adding-deep-immutability/92011/7
-https://discuss.python.org/t/adding-deep-immutability/92011/57
-
-Cashing
-=======
+Deep immutability opens up the possibility of any freezable object being
+hashable, due to its pointer being a stable, unique integer associated
+with the object for its entire post-freeze lifetime. However, there are
+several complications which should be kept in mind for any future PEPs
+which build on this work at add this functionality.
+
+Instance versus Type Hashability
+--------------------------------
+
+At the moment, the test for
+`hashability <https://docs.python.org/3/glossary.html#term-hashable)`_
+is based upon the presence (or absence) of a `__hash__` method and an
+`__eq__` method. Places where `PyObject_HashNotImplemented` is currently
+used would need to be modified as appropriate to have a contextual logic
+which provides a default implementation that uses `id()` if the object
+instance has been frozen, and throws a type error if not.
+
+This causes issues with type checks, however. The check of
+`isinstance(x, Hashable)` would need to become contextual, and 
+`issubclass(type(x), Hashable)` would become underdetermined for
+many types. Handling this in a way that is not surprising will require
+careful design considerations.
+
+Equality of Immutable Objects
+-----------------------------
 
-TODO
-https://discuss.python.org/t/adding-deep-immutability/92011/78
+One consideration with the naive approach (*i.e.*, hash via `id()`) is
+that it can result in confusing outcomes. For example, if there were
+to be two lists:
 
+.. code-block:: python
+  a = [1, 2, 3, 4]
+  b = [1, 2, 3, 4]
+  assert(hash(a) == hash(b))
+
+There would be a reasonable expectation that this assertion would be true,
+as it is for two identically defined tuples. However, without a careful
+implementation of `__hash__` and `__eq__` this would not be the case.
+Our opinion is that an approach like that used in `tuplehash` is
+recommended in order to avoid this behavior.
+
+Decorators of Immutable Functions
+=================================
+
+One natural issue that arises from deeply immutable functions is the
+state of various objects which are attached to them, such as decorators.
+In particular, the case of `lru_cache` is worth investigating. If the cache
+is made immutable, then freezing the function has essentially disabled the
+functionality of the decorator. This might be the correct and desirable
+functionality, from a thread safety perspective! In practice, we see three
+potential approaches:
+
+1. The cache is frozen in its state at the point when freeze is called.
+Cache misses will result in an immutability exception.
+2. Access to the cache is protected by a lock to ensure thread safety
+3. There is one version of the cache per interpreter
```

Comment:
Do we also mention threads to be compatible with NoGil?

```suggestion
3. There is one version of the cache per interpreter or thread
```

Reply:
Slight change in the wording, but good point.

```suggestion
3. There is one version of the cache per interpreter (*i.e.*, the cache is thread local)
```